telemetry/tracks/class-tracks-event.php

```diff
@@ -0,0 +1,336 @@
+
+		$this->data = $validation_result ?? $event_data;
+	}
+
+	/**
+	 * Returns the event's data.
+	 *
+	 * @return Tracks_Event_DTO|WP_Error Event object if the event was created successfully, WP_Error otherwise.
+	 */
+	public function get_data() {
+		return $this->data;
+	}
+
+	/**
+	 * Returns the event's data for JSON representation.
+	 */
+	#[\ReturnTypeWillChange]
+	public function jsonSerialize() {
+		if ( is_wp_error( $this->data ) ) {
+			return (object) [];
+		}
+
+		return $this->data;
+	}
+
+	/**
+	 * Returns whether the event can be recorded.
+	 *
+	 * @return bool|WP_Error True if the event is recordable.
+	 */
+	public function is_recordable() {
+		if ( is_wp_error( $this->data ) ) {
+			return $this->data;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Processes the event's properties to get them ready for validation.
+	 *
+	 * @param string $event_prefix The event's prefix.
+	 * @param string $event_name The event's name.
+	 * @param array<string, mixed>|array<empty> $event_properties Any event properties to be processed.
+	 * @return Tracks_Event_DTO The resulting event object with processed properties.
+	 */
+	protected static function process_properties(
+		string $event_prefix,
+		string $event_name,
+		array $event_properties
+	): Tracks_Event_DTO {
+		$event = static::encode_properties( $event_properties );
+		$event = static::set_user_properties( $event );
+
+		// Set event name. If the event name doesn't have the prefix, add it.
+		$event->_en = preg_replace(
+			'/^(?:' . $event_prefix . ')?(.+)/',
+			$event_prefix . '\1',
+			$event_name
+		) ?? '';
+
+		// Set event timestamp.
+		if ( ! isset( $event->_ts ) ) {
+			$event->_ts = static::milliseconds_since_epoch();
+		}
+
+		// Remove non-routable IPs to prevent record from being discarded.
+		if ( isset( $event->_via_ip ) &&
+			1 === preg_match( '/^192\.168|^10\./', $event->_via_ip ) ) {
+			unset( $event->_via_ip );
+		}
+
+		// Set VIP environment if it exists.
+		if ( defined( 'VIP_GO_APP_ENVIRONMENT' ) ) {
+			$app_environment = constant( 'VIP_GO_APP_ENVIRONMENT' );
+			if ( is_string( $app_environment ) && '' !== $app_environment ) {
+				$event->vipgo_env = $app_environment;
+			}
+		}
+
+		// Set VIP organization if it exists.
+		if ( defined( 'VIP_ORG_ID' ) ) {
+			$org_id = constant( 'VIP_ORG_ID' );
+			if ( is_string( $org_id ) && '' !== $org_id ) {
+				$event->vipgo_org = $org_id;
+			}
+		}
+
+		// Check if the user is a VIP user.
+		$event->is_vip_user = Support_User::user_has_vip_support_role( get_current_user_id() );
+
+		return $event;
+	}
+
+	/**
+	 * Sets the Tracks User ID and User ID Type depending on the current
+	 * environment.
+	 *
+	 * @param Tracks_Event_DTO $event The event to annotate with identity information.
+	 * @return Tracks_Event_DTO The new event object including identity information.
+	 */
+	protected static function set_user_properties( Tracks_Event_DTO $event ): Tracks_Event_DTO {
+		$wp_user = wp_get_current_user();
```

Comment:
This is unsafe, depending on the context the `wp_get_current_user` may not be present. E.g. this would fatal:

```php
add_action( 'muplugins_loaded', fn() => var_dump( wp_get_current_user() ) );
```

Reply:
I've seen this fatal during testing. How would you like to proceed here?

I'm thinking of moving all validation to the write time at shutdown. Any objections?

Reply:
Did that in 275bddd616872420ee28e763c6a0f17c70d3c8e0; please let me know what you think is best.

Reply:
@sjinks Can you please have another look to see if this approach has a different fault? It is now deferred: we are not calling `wp_get_current_user()` from the constructor.